### PR TITLE
docs: clarify max_seq_len numerical behavior

### LINKED ISSUE #4647: trtllm_batch_decode_with_kv_cache: output depends on the host `max_seq_len` argument, breaking eager vs full-CUDA-graph bit-exactness
### Before submitting

- [x] I have searched existing issues and did not find a solution or an existing report.

### Environment

```text
### FlashInfer environment report
<!-- generated by `python -m flashinfer.collect_env`; paste into your issue -->

==== FlashInfer ====
flashinfer                   : 0.6.18
flashinfer commit            : 76704c45003cabaa832d59896080f91dca23f74b
flashinfer file              : /data/users/larryliu/flashinfer/flashinfer/__init__.py
flashinfer-cubin             : not installed
flashinfer-jit-cache         : not installed
JIT cache                    : /home/larryliu/.cache/flashinfer (1 compiled .so)
Cubin store (local)          : /home/larryliu/.cache/flashinfer/cubins (10 cubins, 1 MiB)
Target CUDA archs (resolved) : {(10, '0a')}

==== Python / Platform ====
Python            : 3.12.13+meta (3.12:3bb231a, Mar 03 2026, 12:38:32) [Clang 21.1.0 (ssh://git.vip.facebook.com/data/gitrepos/osmeta/external/llvm-pro
Python executable : /data/users/larryliu/flashinfer/.venv/bin/python
Virtual env       : /data/users/larryliu/flashinfer/.venv
Platform          : Linux-6.16.1-0_fbk3_rc9_0_gd5898487afe6-aarch64-with-glibc2.40
libc              : glibc 2.40
OS                : CentOS Stream 9
Container         : no / not detected

==== GPU / Driver ====
Driver version       : 580.126.20
CUDA_VISIBLE_DEVICES : <unset>
GPU 0 (CUDA order)   : NVIDIA GB200 | SM100 | 152 SMs | 184.3 GiB
GPU 1 (CUDA order)   : NVIDIA GB200 | SM100 | 152 SMs | 184.3 GiB

==== CUDA Toolkit ====
nvcc on PATH                            : not found
CUDA_HOME (env)                         : /usr/local/cuda-13.0
CUDA toolkit resolved by flashinfer JIT : /usr/local/cuda-13.0 (version 13.0)
nvcc version                            : release 13.0, V13.0.88

==== PyTorch ====
torch                          : 2.13.0+cu130
torch.version.cuda             : 13.0
torch CXX11 ABI                : True
torch compiled archs           : sm_80 sm_90 sm_100 sm_110 sm_120
torch.backends.cudnn.version() : 92000
torch file                     : /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/torch/__init__.py

==== cuDNN Frontend ====
cudnn-frontend                       : 1.27.0
cudnn backend (loaded, via frontend) : 9.20.0
cudnn-frontend file                  : /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/cudnn/__init__.py

==== GPU Libraries: loaded vs on disk ====
libcudnn:
    LOADED   /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/nvidia/cudnn/lib/  [v9, pip nvidia-cudnn-cu13==9.20.0.48]
libcublas  ⚠ other installs on disk (check which one you expect):
    LOADED   /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/nvidia/cu13/lib/  [v13, pip nvidia-cublas==13.1.1.3]
    on disk  /usr/local/cuda-12.8/targets/sbsa-linux/lib/  [v12.8.5.5]
    on disk  /usr/local/cuda-13.0/targets/sbsa-linux/lib/  [v13.1.1.3]
libcudart  ⚠ other installs on disk (check which one you expect):
    LOADED   /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/nvidia/cu13/lib/  [v13, pip nvidia-cuda-runtime==13.0.96]
    on disk  /usr/local/cuda-12.8/targets/sbsa-linux/lib/  [v12.8.90]
    on disk  /usr/local/cuda-13.0/targets/sbsa-linux/lib/  [v13.0.96]
libnvrtc  ⚠ other installs on disk (check which one you expect):
    LOADED   /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/nvidia/cu13/lib/  [v13, pip nvidia-cuda-nvrtc==13.0.88]
    on disk  /usr/local/cuda-12.8/targets/sbsa-linux/lib/  [v12.8.93]
    on disk  /usr/local/cuda-13.0/targets/sbsa-linux/lib/  [v13.0.88]
libnccl:
    LOADED   /data/users/larryliu/flashinfer/.venv/lib/python3.12/site-packages/nvidia/nccl/lib/  [v2, pip nvidia-nccl-cu13==2.29.7]
libcuda (driver)  ⚠ other installs on disk (check which one you expect):
    LOADED   /usr/local/fbcode/platform010-aarch64/lib/  [v580.126.20]
    on disk  /usr/lib64/  [v580.126.20]

==== Relevant Packages ====
apache-tvm-ffi               : 0.1.13.post3
cuda-bindings                : 13.3.1
cuda-core                    : 1.0.1
cuda-pathfinder              : 1.6.1
cuda-python                  : 13.3.1
cuda-tile                    : 1.5.0
cuda-toolkit                 : 13.0.3.0
flashinfer-python            : 0.6.17
ninja                        : 1.13.0
numpy                        : 2.5.2
nvidia-cublas                : 13.1.1.3
nvidia-cuda-crt              : 13.3.73
nvidia-cuda-cupti            : 13.0.85
nvidia-cuda-nvcc             : 13.3.73
nvidia-cuda-nvdisasm         : 13.3.73
nvidia-cuda-nvrtc            : 13.0.88
nvidia-cuda-runtime          : 13.0.96
nvidia-cuda-tileiras         : 13.3.36
nvidia-cudnn-cu13            : 9.20.0.48
nvidia-cudnn-frontend        : 1.27.0
nvidia-cufft                 : 12.0.0.61
nvidia-cufile                : 1.15.1.6
nvidia-curand                : 10.4.0.35
nvidia-cusolver              : 12.0.4.66
nvidia-cusparse              : 12.6.3.3
nvidia-cusparselt-cu13       : 0.8.1
nvidia-cutlass-dsl           : 4.7.0
nvidia-cutlass-dsl-libs-base : 4.7.0
nvidia-cutlass-dsl-libs-core : 4.7.0
nvidia-cutlass-dsl-libs-cu12 : 4.7.0
nvidia-cutlass-dsl-libs-cu13 : 4.7.0
nvidia-ml-py                 : 13.610.43
nvidia-nccl-cu13             : 2.29.7
nvidia-nvjitlink             : 13.3.33
nvidia-nvshmem-cu13          : 3.4.5
nvidia-nvtx                  : 13.0.85
nvidia-nvvm                  : 13.3.73
torch                        : 2.13.0+cu130
triton                       : 3.7.1

==== Environment Variables ====
CUDA_CACHE_PATH : /data/users/larryliu/.nv/ComputeCache
CUDA_HOME       : /usr/local/cuda-13.0
CUDA_PATH       : /usr/local/cuda-12.6
LD_LIBRARY_PATH : /usr/local/cuda-13.0/lib64:/usr/local/cuda-12.6/lib64:/usr/local/cuda-12.9/lib64/:/data/users/larryliu/qairt/2.37.0.250724/lib:/usr/local/cuda-13.0/lib64:/usr/local/cuda-12.6/lib64:/usr/local/cuda-12.9/lib64/:/data/users/larryliu/qairt/2.37.0.250724/lib:
PATH            : /data/users/larryliu/flashinfer/.venv/bin:/home/larryliu/.local/bin:/usr/local/cuda-12.6/bin:/usr/local/cuda-12.9/bin:/data/users/larryliu/qairt/2.37.0.250724/bin:/data/users/larryliu/android-ndk-r27d:/usr/local/fbpkg/vscodefb/vscode-server/434/bin/remote-cli:/home/larryliu/.local/bin:/usr/local/cuda-12.6/bin:/usr/local/cuda-12.9/bin:/data/users/larryliu/qairt/2.37.0.250724/bin:/data/users/larryliu/android-ndk-r27d:/usr/condabin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/usr/facebook/ops/scripts:/usr/facebook/scripts:/usr/facebook/scripts/db:/sbin:/var/www/scripts/bin:/home/larryliu/.yarn/bin:/opt/android_sdk/tools:/opt/android_sdk/platform-tools:/home/larryliu/local/idea-IC-173.4548.28/bin:/home/larryliu/bin:/usr/facebook/ops/scripts:/usr/facebook/scripts:/usr/facebook/scripts:/usr/facebook/scripts/db:/usr/facebook/ops/scripts:/usr/facebook/scripts:/usr/local/sbin:/usr/sbin:/sbin:/var/www/scripts/bin:/home/larryliu/.yarn/bin:/opt/android_sdk/tools:/opt/android_sdk/platform-tools:/home/larryliu/local/idea-IC-173.4548.28/bin
VIRTUAL_ENV     : /data/users/larryliu/flashinfer/.venv

==== GPU Topology ====
nvidia-smi topo -m:
GPU0    GPU1    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    SYS     SYS     SYS     SYS     SYS     SYS     0-71    0,2-9           N/A
GPU1    NV18     X      SYS     SYS     SYS     SYS     SYS     SYS     72-143  1,10-17         N/A
NIC0    SYS     SYS      X      SYS     SYS     SYS     SYS     SYS
NIC1    SYS     SYS     SYS      X      SYS     SYS     SYS     SYS
NIC2    SYS     SYS     SYS     SYS      X      SYS     SYS     SYS
NIC3    SYS     SYS     SYS     SYS     SYS      X      SYS     SYS
NIC4    SYS     SYS     SYS     SYS     SYS     SYS      X      SYS
NIC5    SYS     SYS     SYS     SYS     SYS     SYS     SYS      X
```

### Bug description

  `trtllm_batch_decode_with_kv_cache` returns different results when only the host-side
  `max_seq_len` argument changes. Every tensor input — including the device `seq_lens`
  buffer — is byte-identical; only the host int differs.

  `max_seq_len` is documented as *"max sequence length for kv_cache"*, i.e. an upper
  bound, and no validation ties it to `max(seq_lens)`. This matters because it is a
  **host** int: under full CUDA-graph capture it is frozen at capture time, so a serving
  stack (e.g. vLLM with `cudagraph_mode=FULL_AND_PIECEWISE`) has no choice but to bake in
  `max_model_len` and can never reproduce its own eager/piecewise output bit-for-bit.

  **This is not an accuracy regression.** Against an fp32 reference both paths are
  equidistant and both stay under 1 bf16 ULP. The problems are (a) non-reproducibility
  across an argument documented as a bound, and (b) wasted work — a 253-token sequence
  gets a 19-way KV split plus a gmem reduction.

  Root cause appears to be that `mMaxSeqLenKv` drives two host-side heuristics:

  - `include/flashinfer/trtllm/fmha/fmhaKernels.cuh:617` computes `maxNumCtasPerSeqKv`
    from `mMaxSeqLenKv`, and `:649` derives the KV-split factor `numCtasPerSeqKv`.
    `:657` disables multi-CTA mode (and re-selects a kernel with the persistent
    scheduler) when the split is 1; `:679` upgrades to `CgaSmemReduction` when it is ≤ 16.
  - `:1051-1060` uses `mMaxSeqLenKv` in the kernel-selection cost model that picks
    `tileSizeQ`.

  A different split / reduction mode / kernel changes the accumulation order, hence the
  bf16 rounding. The cliff is visible at `seq_len` 4864 → 4865, which is exactly where
  `ceil(seq_len / 256)` crosses the SM-derived cap on a 152-SM GB200.

  The effect disappears at large batch (batch=32 shows bit-equality), because the split
  heuristic never engages there.

  ### Expected

  For a fixed set of tensor inputs, any `max_seq_len >= max(seq_lens)` should give the
  same result — or the docstring should state that `max_seq_len` perturbs numerics and
  that eager/CUDA-graph bit-exactness is not available.

  ### Output

  ```
  NVIDIA GB200  batch=1, q_len=1, 64q/8kv heads, head_dim=64, page=16, bf16

  seq_lens=[253]
     max_seq_len  bit-equal  max|d vs exact-hint|  max|d vs fp32|
             253       True             0.000e+00       9.863e-04
            8192      False             1.953e-03       1.815e-03
           32768      False             1.953e-03       1.815e-03
           65535      False             1.953e-03       1.815e-03

  seq_lens=[1024]
     max_seq_len  bit-equal  max|d vs exact-hint|  max|d vs fp32|
            1024       True             0.000e+00       9.955e-04
            8192      False             9.766e-04       8.219e-04
           32768      False             9.766e-04       8.219e-04
           65535      False             9.766e-04       8.219e-04

  seq_lens=[4096]
     max_seq_len  bit-equal  max|d vs exact-hint|  max|d vs fp32|
            4096       True             0.000e+00       3.865e-04
            8192      False             4.883e-04       3.251e-04
           32768      False             4.883e-04       3.251e-04
           65535      False             4.883e-04       3.251e-04

  seq_lens=[4864]
     max_seq_len  bit-equal  max|d vs exact-hint|  max|d vs fp32|
            4864       True             0.000e+00       3.752e-04
            8192       True             0.000e+00       3.752e-04
           32768      False             4.883e-04       3.565e-04
           65535      False             4.883e-04       3.565e-04

  seq_lens=[4865]
     max_seq_len  bit-equal  max|d vs exact-hint|  max|d vs fp32|
            4865       True             0.000e+00       3.077e-04
            8192       True             0.000e+00       3.077e-04
           32768       True             0.000e+00       3.077e-04
           65535       True             0.000e+00       3.077e-04
  ```

### Minimal reproduction

```python
"""trtllm_batch_decode_with_kv_cache: the output changes when only the host-side
  `max_seq_len` argument changes."""

  import torch
  from flashinfer.decode import trtllm_batch_decode_with_kv_cache

  H_Q, H_KV, D, P = 64, 8, 64, 16
  DEV = "cuda"


  def fp32_reference(q, kv_cache, block_table, seq_len):
      pages = block_table[: (seq_len + P - 1) // P]
      k = kv_cache[pages, 0].permute(1, 0, 2, 3).reshape(H_KV, -1, D)[:, :seq_len].float()
      v = kv_cache[pages, 1].permute(1, 0, 2, 3).reshape(H_KV, -1, D)[:, :seq_len].float()
      rep = H_Q // H_KV
      k, v = k.repeat_interleave(rep, 0), v.repeat_interleave(rep, 0)
      s = (q.float().squeeze(0).unsqueeze(1) @ k.transpose(1, 2)) * D**-0.5
      return (s.softmax(-1) @ v).squeeze(1).unsqueeze(0)


  def main():
      torch.manual_seed(0)
      n_pages = 65536 // P
      # canonical HND kv cache: [num_pages, 2, num_kv_heads, page_size, head_dim]
      kv_cache = torch.randn(n_pages, 2, H_KV, P, D, device=DEV, dtype=torch.bfloat16)
      block_table = torch.arange(n_pages, dtype=torch.int32, device=DEV)
      workspace = torch.zeros(128 * 1024 * 1024, dtype=torch.uint8, device=DEV)
      q = torch.randn(1, H_Q, D, device=DEV, dtype=torch.bfloat16)

      def decode(seq_lens, max_seq_len):
          return trtllm_batch_decode_with_kv_cache(
              query=q,
              kv_cache=kv_cache,
              workspace_buffer=workspace,
              block_tables=block_table.unsqueeze(0),
              seq_lens=seq_lens,
              max_seq_len=max_seq_len,
              bmm1_scale=D**-0.5,
              bmm2_scale=1.0,
              backend="trtllm-gen",
          ).clone()

      print(f"{torch.cuda.get_device_name(0)}  batch=1, q_len=1, {H_Q}q/{H_KV}kv heads, "
            f"head_dim={D}, page={P}, bf16\n")

      for seq_len in (253, 1024, 4096, 4864, 4865):
          seq_lens = torch.tensor([seq_len], dtype=torch.int32, device=DEV)
          ref = fp32_reference(q, kv_cache, block_table, seq_len)
          exact = decode(seq_lens, seq_len)
          print(f"seq_lens=[{seq_len}]")
          print(f"  {'max_seq_len':>12} {'bit-equal':>10} {'max|d vs exact-hint|':>21}"
                f" {'max|d vs fp32|':>15}")
          for hint in (seq_len, 8192, 32768, 65535):
              if hint < seq_len:
                  continue
              out = decode(seq_lens, hint)
              print(f"  {hint:>12} {str(torch.equal(out, exact)):>10}"
                    f" {(out.float() - exact.float()).abs().max().item():>21.3e}"
                    f" {(out.float() - ref).abs().max().item():>15.3e}")
          print()


  if __name__ == "__main__":
      main()
```

### Error messages / logs

```text

```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by XFDG**
## Summary

- clarify that max_seq_len is a host-side upper bound for KV sequence lengths
- document that the TRT-LLM backend uses this hint for KV split, reduction, and kernel selection
- explain the resulting floating-point reproducibility contract and how to keep eager and CUDA Graph execution bitwise aligned

## Testing

- ruff format --check flashinfer/decode.py
- ruff check flashinfer/decode.py
- git diff --check
- reproduced the reported hint-dependent BF16 output on B200/SM100 with current main; all paths remained within expected numerical tolerance

Fixes #4647

## Summary by CodeRabbit

* **Documentation**
  * Expanded `max_seq_len` documentation for batch decoding with a KV cache on `trtllm-gen` devices with compute capability major 10.
  * Clarified its effects on tile size, KV splitting, reduction strategy, extra work, accumulation order, and numerical equivalence.
  * Documented the requirement to use the same value during eager execution and CUDA Graph capture for bitwise reproducibility.

### CODE REVIEW

File: flashinfer/decode.py [hunk 1]
```diff
@@ -3515,7 +3515,17 @@ def trtllm_batch_decode_with_kv_cache(
...
     max_seq_len : int
-        max sequence length for kv_cache
+        Host-side upper bound on the sequence lengths in ``kv_cache``. It must be at
+        least ``max(seq_lens)``.
+
+        For the ``trtllm-gen`` backend, this value participates in kernel selection,
+        including the KV split and reduction strategy. Consequently, different valid
+        upper bounds can change the floating-point accumulation order, and their
+        outputs are not guaranteed to be bitwise identical (although they remain
+        numerically equivalent within the documented dtype tolerances). Because CUDA
```
[the remark this answers is not in the document]
Comment (XFDG) [reply flashinfer/decode.py]: Addressed in 28946cb7. The contract now states that, on supported compute-capability-major-10 devices, different valid upper bounds may change tile sizes and introduce extra work, in addition to changing KV splitting, reduction, and accumulation order.

[the remark this answers is not in the document]
Comment (XFDG) [reply flashinfer/decode.py]: Addressed in 28946cb7. I replaced the broad documented-dtype-tolerances claim with a qualified statement that numerical agreement is limited to the tolerances exercised by the supported SM10x test matrix.
